File: src/data/courses/dsa/string-hashing-problems/hashing-based-problems/product-of-array-except-self/code/cpp/product_of_array_except_self_sol3.cpp
class Solution {
public:
    // Optimal: O(n) time, O(1) extra space — Two passes with running variable
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n, 1);
        int leftProduct = 1;
        for (int i = 0; i < n; i++) {       // Pass 1: left products
            result[i] = leftProduct;
            leftProduct *= nums[i];
        }
        int rightProduct = 1;
        for (int i = n-1; i >= 0; i--) {    // Pass 2: multiply by right products
            result[i] *= rightProduct;
            rightProduct *= nums[i];
        }
        return result;
    }
};

File: src/data/courses/dsa/string-hashing-problems/hashing-based-problems/product-of-array-except-self/code/python/product_of_array_except_self_sol3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # Optimal: O(n) time, O(1) extra space
        n = len(nums)
        result = [1] * n
        left = 1
        for i in range(n):
            result[i] = left
            left *= nums[i]
        right = 1
        for i in range(n-1, -1, -1):
            result[i] *= right
            right *= nums[i]
        return result